internal/karma/ts_web_test.bzl

```diff
@@ -78,6 +78,15 @@ def _ts_web_test_impl(ctx):
         for f in ctx.files.static_files
     ]
 
+    setup_files = []
+    for d in ctx.attr.setup_deps:
+        if not hasattr(d, "typescript"):
```

Comment:
You're checking if the setup_dep has a typescript provider so that you can get at the es5_sources? Seems like an overly strict requirement since it prevents a user from adding a .js file to setup_files. Maybe this should follow the same pattern as user_entries:
```
setup_deps = depset([])
for d in ctx.attr.setup_deps:
    if hasattr(d, "node_sources"):
        setup_deps += d.node_sources
    elif hasattr(d, "files"):
        setup_deps += d.files
setup_files = [
    expand_path_into_runfiles(ctx, f.short_path)
    for f in setup_deps
]
```
The above doesn't check for the single file requirement but that could be done as well in the loop if it is desirable

Reply:
I think it makes sense to also allow js_library files, but I don't think it will work with generic JS files since trying to `require()` them will fail.

Is it more idiomatic in bazel to guard against unexpected inputs, or to be permissive?